### ISSUE TITLE
ThinLTO strips symbols defined in inline assembly

### ISSUE THREAD

**Issue by amyspark** [llvm/llvm-project#64127]
Using Clang 16.0.5 to compile [this fork of FFmpeg](https://gitlab.freedesktop.org/seungha.yang/ffmpeg/-/tree/win-static-build) returns the following error:

```console
[3721/4957] Linking target subprojects/FFmpeg/test_avcodec_avcodec.exe
FAILED: subprojects/FFmpeg/test_avcodec_avcodec.exe
"cc"  -o subprojects/FFmpeg/test_avcodec_avcodec.exe subprojects/FFmpeg/test_avcodec_avcodec.exe.p/libavcodec_tests_avcodec.c.obj "-flto=thin" "-Wl,--start-group" "subprojects\FFmpeg\libavcodec-static.a" "subprojects\FFmpeg\libavutil.a" "subprojects\FFmpeg\libavutil-static.a" "subprojects\FFmpeg\libswresample.a" "subprojects\FFmpeg\libswresample-static.a" "-pthread" "-lws2_32" "-lm" "-lbcrypt" "-lz" "-llzma" "-lws2_32" "-lbcrypt" "-lz" "-llzma" "-lws2_32" "-lbcrypt" "-lws2_32" "-lbcrypt" "-lws2_32" "-lbcrypt" "-lws2_32" "-lbcrypt" "-Wl,--subsystem,console" "-lkernel32" "-luser32" "-lgdi32" "-lwinspool" "-lshell32" "-lole32" "-loleaut32" "-luuid" "-lcomdlg32" "-ladvapi32" "-Wl,--end-group"
ld.lld: error: undefined symbol: ff_mlp_firorder_0
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_1
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_2
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_3
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_4
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_5
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_6
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_7
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_firorder_8
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_iirorder_0
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_iirorder_1
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_iirorder_2
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_iirorder_3
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj

ld.lld: error: undefined symbol: ff_mlp_iirorder_4
>>> referenced by ../subprojects/FFmpeg/libavcodec/x86/mlpdsp_init.c
>>>               subprojects\FFmpeg/libavcodec-static.a.p/libavcodec_x86_mlpdsp_init.c.obj
```

It looks like the same bug as https://github.com/ziglang/zig/issues/10364, which was never reported here. The issue seems to be that ThinLTO doesn't know that those `extern` symbols are in fact defined inside the chunk of inline assembly.

Example Godbolt (see where the symbols are defined): https://godbolt.org/z/eqrd5WG8x

To reproduce: under a MSYS2 CLANG64 shell, run:
```sh
meson setup build -Dgst-plugins-base:pango=disabled -Dpython=disabled -Dgst-plugins-good:png=disabled -Ddevtools=disabled -Ddefault_library=static -Dgst-full-target-type=static_library -Db_lto=true -Db_lto_mode=thin -Dprefix=C:/Work/static_install
meson compile -C build
```

**Comment by amyspark** [comment]
> To avoid this issue, symbols used in inline assembly should be listed as an input operand on the inline asm call

@teresajohnson In this case, it's the inline asm call that "defines" those symbols, which are consumed elsewhere: https://github.com/FFmpeg/FFmpeg/blob/0ba719f726632d73592311615087a0d64aa2fb60/libavcodec/x86/mlpdsp_init.c#L50-L64

> or alternatively can be annotated with __attribute__((used))

It cannot be used, as it can only affect real C-side definitions.

By the way, this was tested with Clang on MSYS2, I have not checked if it affects Microsoft-shipped Clang versions.

**Comment by teresajohnson** [comment]
Not an inline asm expert, but since these variables are defined in inline asm, can't/shouldn't they then be made output operands? 

Generally these are opaque blobs to compilers (not just clang, gcc has similar treatment: https://gcc.gnu.org/onlinedocs/gcc/extensions-to-the-c-language-family/how-to-use-inline-assembly-language-in-c-code.html#how-to-use-inline-assembly-language-in-c-code).

**Comment by amyspark** [comment]
No, because of https://github.com/ffmpeg/ffmpeg/commit/c8252e80eb1d0d0964e94b3186bfdb776fee99d6. They're not real outputs.

Also update: with Microsoft-shipped Clang this does not occur, probably because it's a ld-lld ThinLTO semantic.

**Comment by teresajohnson** [comment]
I'm not really sure how output labels are supposed to be specified to the rest of the compiler. I'm not finding a lot online about this situation, only about labels local to the inline assembly.

> Also update: with Microsoft-shipped Clang this does not occur, probably because it's a ld-lld ThinLTO semantic.

For Microsoft-shipped Clang, is that with any type of LTO? lld and ThinLTO only know about what is in the LLVM IR, generated by Clang, and as mentioned, inline asm is AFAIK opaque (other than the specified input/output operands and the like).

**Comment by teresajohnson** [comment]
After thinking about this some more, here's what I suspect is happening:

In a non-LTO compile, the linker only sees native objects, and at that point the code has gone all the way through the code generator and integrated assembler, so the inline asm is assembled and the label is defined in the native object.

In an LTO compile (either regular or ThinLTO), there are essentially 2 links happening:
1) the LTO link, where the linker does its initial symbol resolution, relying on LLVM to tell it what symbols are defined and used in each LTO bitcode module.
2) The final native link using the native objects LLVM's LTO passes back to the linker after going all the way through the compilation pipeline including the final code gen / assembler.

I believe what is happening is that the undefined reference errors are coming from the symbol resolution during step 1 above. As I noted earlier, inline asm is an opaque blob in LLVM IR, so it has no way of knowing that the label is defined there. 

For symbols defined in normal source code but used in the inline asm, we rely on attribute(used) annotations so that they don't appear dead. But this won't work for telling the compiler what labels are defined in the inline asm. I'm not sure what the solution is, but since we compile ffmpeg with ThinLTO internally I will take a look and see why we aren't getting this error there.

**Comment by teresajohnson** [comment]
Hmm, passing -Wl,-y,ff_mlp_firorder_0 I can see that there is a reference from the LTO object for that source file, but no definition is reported anywhere during the LTO link. However, there is no undefined reference error. I'm not really sure why, maybe during the LTO link this is not a hard error? @MaskRay to comment on lld behavior.

Also, my testing was with ELF lld, I'm not sure whether the lld COFF LTO behavior could be different.

**Comment by amyspark** [comment]
I just tried with AppleClang (1300.0.29.30) and it seems that both ELF lld and MachO (AppleClang) lld can deal with inline assembly generated symbols. Only COFF lld is stripping them.

**Comment by amyspark** [comment]
Wondering if https://github.com/llvm/llvm-project/blob/bbe2887f5e9ca005b0f1b96c858969ee3ba646f4/lld/COFF/LTO.cpp#L149-L157 has something to do with this bug.

**Comment by teresajohnson** [comment]
> I just tried with AppleClang (1300.0.29.30) and it seems that both ELF lld and MachO (AppleClang) lld can deal with inline assembly generated symbols. Only COFF lld is stripping them.

@amykhuang 
@ZequanWu 

Thanks for confirming. Someone who has worked on COFF LTO in lld should probably look (referenced a couple of folks above). As noted in one of my earlier responses, I suspect this isn't the symbol being stripped, but rather the initial LTO link treating the symbol as undefined since we don't parse inline asm instructions at that point and don't see the definition.

> Wondering if https://github.com/llvm/llvm-project/blob/bbe2887f5e9ca005b0f1b96c858969ee3ba646f4/lld/COFF/LTO.cpp#L149-L157 has something to do with this bug.

That same code with the same comment exists in ELF lld as well. It also relates to module asm, which is different than inline asm calls (and module asm does get parsed by LLVM so are treated differently).

**Comment by amyspark** [comment]
Hi again, were you able to look into this? I may have an idea why it's failing, but I've not had the opportunity to try and build LLVM for Windows from source.

**Comment by amyspark** [comment]
@teresajohnson Here: https://github.com/amyspark/llvm-bug-64127 and once Matt enables UCRT to run CMake projects: https://godbolt.org/z/oo1bWz4KK

**Comment by MaskRay** [comment]
tl;dr Ensure `ff_mlp_iirorder_*` are defined as regular functions or use lld-link `-force:unresolved`

> @teresajohnson Here: https://github.com/amyspark/llvm-bug-64127 and once Matt enables UCRT to run CMake projects: https://godbolt.org/z/oo1bWz4KK

Thank you for the minimal reproduce. I don't have a MinGW environment but I can reproduce the error.

```
% clang -c --target=x86_64-windows-msvc -nostdlibinc main.cpp io.c -I. -flto
% lld-link -force:unresolved main.o io.o
lld-link: error: <root>: undefined symbol: mainCRTStartup
lld-link: error: undefined symbol: ff_mlp_firorder_8
>>> referenced by io.c
>>>               io.o

ld.lld: error: undefined symbol: ff_mlp_firorder_0
>>> referenced by io.c
>>>               io.o
...
% llvm-nm io.o
                 U ff_mlp_firorder_0
                 U ff_mlp_firorder_1
                 U ff_mlp_firorder_2
                 U ff_mlp_firorder_3
                 U ff_mlp_firorder_4
                 U ff_mlp_firorder_5
                 U ff_mlp_firorder_6
                 U ff_mlp_firorder_7
                 U ff_mlp_firorder_8
                 U ff_mlp_iirorder_0
                 U ff_mlp_iirorder_1
                 U ff_mlp_iirorder_2
                 U ff_mlp_iirorder_3
                 U ff_mlp_iirorder_4
---------------- T ff_mlpdsp_init_x86
---------------- d firtable
---------------- d iirtable
---------------- t mlp_filter_channel_x86
```

`ff_mlp_iirorder_0` is defined in function-local inline assembly. The IR symbol table does not parse function-local inline assembly and does not know there is a definition.
ELF and Mach-O linkers use a model that they only report "undefined symbol"/"undefined reference" errors for referenced undefined symbols.
However, link.exe and lld/COFF report "unresolved external symbol"/"undefined symbol" errors regardless of whether the symbols are referenced.

Extended asm requires operands to be finalized at a later stage, and in the IR it cannot be parsed reliably. Therefore, it's not really easy to make `ff_mlp_iirorder_0` defined.
As a workaround, specify `-force:unresolved`:

```
% lld-link -force:unresolved main.o io.o
lld-link: warning: <root>: undefined symbol: mainCRTStartup
```

A better way is to define `ff_mlp_firorder_0` as a weak definition in a translation unit different from the inline asm definition:
```c
__attribute__((weak)) void ff_mlp_firorder_0() {}
```
